whois-nrtm4/src/main/java/net/ripe/db/nrtm4/NrtmFileService.java

```diff
@@ -0,0 +1,115 @@
+
+    private final NotificationFileRepository notificationFileRepository;
+    private final NrtmFileRepo nrtmFileRepo;
+    private final NrtmVersionInfoRepository nrtmVersionInfoRepository;
+    private final DeltaFileRepository deltaFileRepository;
+    private final SnapshotFileRepository snapshotFileRepository;
+    private final SnapshotFileStreamer snapshotFileStreamer;
+    private final Monitor deltaMutex = new Monitor();
+    private final Monitor snapshotMutex = new Monitor();
+
+    NrtmFileService(
+        final NotificationFileRepository notificationFileRepository,
+        final NrtmFileRepo nrtmFileRepo,
+        final NrtmVersionInfoRepository nrtmVersionInfoRepository,
+        final DeltaFileRepository deltaFileRepository,
+        final SnapshotFileRepository snapshotFileRepository,
+        final SnapshotFileStreamer snapshotFileStreamer
+    ) {
+        this.notificationFileRepository = notificationFileRepository;
+        this.nrtmFileRepo = nrtmFileRepo;
+        this.nrtmVersionInfoRepository = nrtmVersionInfoRepository;
+        this.deltaFileRepository = deltaFileRepository;
+        this.snapshotFileRepository = snapshotFileRepository;
+        this.snapshotFileStreamer = snapshotFileStreamer;
+    }
+
+    void writeFileToStream(final String name, final OutputStream out) throws IOException {
+        if (name == null || name.length() > 256) {
+            throw new IllegalArgumentException("Invalid NRTM file name");
+        }
+        if (name.startsWith(NOTIFICATION_PREFIX)) {
+            // TODO: how do we know which source to use when serving a notification request?
+            final NotificationFile notificationFile = notificationFileRepository.getNotificationFile(1);
+            out.write(notificationFile.getPayload().getBytes(StandardCharsets.UTF_8));
+            return;
+        }
+        syncNrtmFileToFileSystem(name);
+        nrtmFileRepo.streamFromFile(name, out);
+    }
+
+    void syncNrtmFileToFileSystem(final String name) throws IOException {
+        if (name.startsWith(DELTA_PREFIX)) {
+            if (!nrtmFileRepo.checkIfFileExists(name)) {
+                deltaMutex.enter();
+                try {
+                    // check again now that the lock is in force
+                    if (!nrtmFileRepo.checkIfFileExists(name)) {
+                        final Optional<DeltaFile> optDeltaFile = deltaFileRepository.getByName(name);
+                        if (optDeltaFile.isPresent()) {
+                            nrtmFileRepo.storeFile(name, optDeltaFile.get().getPayload());
+                        } else {
+                            throw new FileNotFoundException("NRTM has no delta files with name: " + name);
+                        }
+                    }
+                } finally {
+                    deltaMutex.leave();
+                }
+            }
+        } else if (name.startsWith(SNAPSHOT_PREFIX)) {
+            if (!nrtmFileRepo.checkIfFileExists(name)) {
```

Comment:
You must have got the filename from the nrtm database when you are trying to write it? If yes , I would rather rely on documnetType enum rather than filename.

Reply:
this code is for dealing with incoming client requests -- it checks if the file exists on disk before going to the database.